### PR TITLE
Remove assemble task when not used for publishing

### PR DESCRIPTION
**Pull request by nik9000**
Removes the `assemble` task from projects that are not published.
This should speed up `gradle assemble` by skipping projects that
don't need to be built. Which is useful because `gradle assemble`
is how we cut releases.

### CODE REVIEW

File: benchmarks/build.gradle [hunk 1]
```diff
@@ -34,6 +34,8 @@ apply plugin: 'com.github.johnrengelman.shadow'
 // have the shadow plugin provide the runShadow task
 apply plugin: 'application'
 
+tasks.remove(assemble)  // Not published so no need to assemble
```
Comment (jasontedor) [inline benchmarks/build.gradle]: How do you live with two spaces after between the content and the comment? It 🔥 my 👀. 😛

**Review by jasontedor — APPROVED** [review]
OCD aside, LGTM.

File: benchmarks/build.gradle (hunk 1 above)
Comment (nik9000) [inline benchmarks/build.gradle]: You like one space?

Comment (jasontedor) [inline benchmarks/build.gradle]: Yes!

File: build.gradle [hunk 2]
```diff
@@ -402,3 +402,16 @@ task run(type: Run) {
...
   impliesSubProjects = true
 }
+
+/* Remove assemble on all qa projects because we don't need to publish
+ * artifacts for them. */
+gradle.projectsEvaluated {
+  subprojects {
+    if (project.path.startsWith(':qa')) {
+      Task assemble = project.tasks.findByName('assemble')
+      if (assemble) {
```
Comment (rjernst) [inline build.gradle]: This evaluation should only run once, so why is there a condition necessary? Seems like it should always be there?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy [hunk 3]
```diff
@@ -394,8 +394,11 @@ class BuildPlugin implements Plugin<Project> {
             project.tasks.withType(GenerateMavenPom.class) { GenerateMavenPom t ->
                 // place the pom next to the jar it is for
                 t.destination = new File(project.buildDir, "distributions/${project.archivesBaseName}-${project.version}.pom")
-                // build poms with assemble
-                project.assemble.dependsOn(t)
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: We should only have the pom generation task on projects that will be shipped, so it should always have an assemble task, and this change should not be necessary.

**Review by rjernst — APPROVED** [review]
Looks good, a couple comments.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy (hunk 3 above)
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: I believe plugins get pom generation configured for their zip file whether or not we publish the zip file. I can build a thing to make them not get pom generation, I imagine.

File: build.gradle (hunk 2 above)
Comment (nik9000) [inline build.gradle]: `:qa:vagrant` doesn't have an `assemble` task but the rest do.

**Pull request merged**